_server.js

```diff
@@ -0,0 +1,158 @@
+const electron = require('electron'),
+app = electron.app,
+BrowserWindow = electron.BrowserWindow,
+ipcMain = electron.ipcMain
+
+global.directory = 'file://' + __dirname + '/'
+
+const shadyServer = require('express')(),
+shadowServer = require('express')(),
+shadyPort = 3000,
+shadowPort = 4000,
+path = require('path'),
+ngrok = require('ngrok')
+
+const fs = require('fs'),
+url = require('url')
+
+let win
+
+var listening = false,
+shadyRes = null,
+shadowRes = null,
+shadyAsyncImports = '',
+shadowAsyncImports = ''
+
+const LRUCache = require('./lruCache'),
+  cacheSize = 5, // arbitrary number
+  cache = new LRUCache(cacheSize),
+  hasher = require('string-hash')
+
+// Setup for Electron app
+function createWindow() {
+  win = new BrowserWindow({width: 800, height: 600})
+  win.loadURL('data:text/html;charset:utf-8,<html></html>')
+  win.webContents.openDevTools()
+
+  win.on('closed', () => {
+    win = null
+  })
+
+  // Setup for Express server
+  win.webContents.on('dom-ready', () => {
+    startServer();
+  })
+}
+
+app.on('ready', createWindow)
+
+app.on('window-all-closed', () => {
+  if (process.platform !== 'darwin') {
+    stopServer()
+    app.quit()
+  }
+})
+
+app.on('activate', function () {
+  if (mainWindow === null) {
+    createWindow()
+  }
+})
+
+// IPC functions
+ipcMain.on('receiveSerializedDOM', (_, contents, isShady) => {
+  if(isShady) {
+    shadyRes.end(contents)
+  } else {
+    shadowRes.end(contents)
+  }
+})
+
+ipcMain.on('setAsyncImports', (event, key, value) => {
+  cache.set('_asyncImport' + key + '.html', value)
+  event.returnValue = true;
+})
+
+shadyServer.get(/\/index[0-9]*.html/, (req, res) => {
+  win.loadURL('file://' + __dirname + req.url)
+  shadyRes = res
+
+  win.webContents.on('did-finish-load', () => {
+    shadyGetDOMInsidePage()
+  })
+})
+
+shadowServer.get(/\/index[0-9]*shadow.html/, (req, res) => {
+  win.loadURL('file://' + __dirname + req.url)
+  shadowRes = res
+
+  win.webContents.on('did-finish-load', () => {
+    // shadyThenShadowGetDOMInsidePage()
+    shadowGetDOMInsidePage()
+  })
+})
+
+shadyServer.get(/_asyncImport[0-9]+/, (req, res) => {
+  getAsyncImport(req.url.substring(1), res)
+})
+
+shadowServer.get(/_asyncImport[0-9]+/, (req, res) => {
+  getAsyncImport(req.url.substring(1), res)
+})
+
+function getAsyncImport(key, res) {
+  res.end('' + cache.get(key))
+}
```

Comment:
`getAsyncImport` can be the one request handler instead of having two request handlers calling it:

```js
shadyServer.get(/_asyncImport[0-9]+/, getAsyncImport);
shadowServer.get(/_asyncImport[0-9]+/, getAsyncImport);
 
function getAsyncImport(req, res) {
  res.end('' + cache.get(req.url.substring(1)));
}
```